pkg/controllers/scheduling/controller.go

```diff
@@ -0,0 +1,178 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/awslabs/karpenter/pkg/controllers/provisioning"
+	"github.com/awslabs/karpenter/pkg/utils/pod"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient   client.Client
+	provisioners *sync.Map
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, provisioners *sync.Map) *Controller {
+	return &Controller{
+		kubeClient:   kubeClient,
+		provisioners: provisioners,
+	}
+}
+
+// Reconcile the resource
+func (r *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("scheduling").With("pod", req.String()))
+	pod := &v1.Pod{}
+	if err := r.kubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Ensure the pod can be provisioned, but only log
+	if err := isUnschedulable(pod); err != nil {
+		return reconcile.Result{}, nil
+	}
+	if err := validate(pod); err != nil {
+		logging.FromContext(ctx).Debugf("Ignoring pod, %s", err.Error())
+		return reconcile.Result{}, nil
+	}
+	// Schedule and requeue. If successful, will terminate in the unschedulable check above
+	if err := r.Schedule(ctx, pod); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to schedule, %s", err.Error())
+	}
+	return reconcile.Result{Requeue: true}, nil
+}
+
+func (r *Controller) Schedule(ctx context.Context, pod *v1.Pod) (error) {
+	var provisioner *provisioning.Provisioner
+	var errs error
+	r.provisioners.Range(func(_, p interface{}) bool {
+		candidate := p.(*provisioning.Provisioner)
+		if err := candidate.Spec.Supports(pod); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tried provisioner/%s: %w", candidate.Name, err))
+		} else {
+			provisioner = candidate
+			return false
+		}
+		return true
+	})
+	if provisioner == nil {
+		err := fmt.Errorf("matched 0/%d provisioners", len(multierr.Errors(errs)))
+		if errs != nil {
+			err = fmt.Errorf("%s, %w", err, errs)
+		}
+		return err
+	}
+	logging.FromContext(ctx).Debugf("Matched provisioner %q", provisioner.Name)
+	if err := provisioner.Enqueue(ctx, pod); err != nil {
+		return fmt.Errorf("provisioner %q failed", provisioner.Name)
```

Comment:
I think the error is needed here too?

Reply:
I trimmed it since its hugely verbose. The behavior is that the top level provisioner will fail, and then each of the pods that relied on it will say that it also failed.